### ISSUE TITLE
Link does not obey Router basepath

### ISSUE THREAD

**Issue by Pyrolistical** [reach/router#78]
Try this demo: https://pyrolistical.github.io/reach-router-basepath-demo/

Dashboard link is fine, Home link breaks.

I've set the basepath here: https://github.com/Pyrolistical/reach-router-basepath-demo/blob/master/src/index.js#L29

The Home link is `<Link to="/">` and changes the history to be `https://pyrolistical.github.io/` when it should have set it to `https://pyrolistical.github.io/reach-router-basepath-demo/`

Full source: https://github.com/Pyrolistical/reach-router-basepath-demo

**Comment by Pyrolistical** [comment]
`navigate` directly comes from `src/lib/history.js`. I don't see any reason why an app would want to change history outside of where it is deployed. We could pass `basepath` into `navigate`, so it knows it needs to include it when changing the raw history.

**Comment by Pyrolistical** [comment]
The workaround is to include the basepath on all absolute paths in `Link` and `navigate`

**Comment by alexkrolick** [comment]
I proposed #75 to work around this, but I guess it's intentional to be more directly in line with browser behavior

**Comment by timdorr** [comment]
`basepath` is equivalent to `<base href="...">` in HTML. And [a definition of it](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base) is this:

> The base URL to be used throughout the document for relative URL addresses. 

In the router world, since relative URLs are already handled anyways, it doesn't really factor into links. It's for matching incoming URLs from the browser, so you can mount your app at any sub-path and still have it work. 

If you have relative URLs everywhere, you'll be fine. Understandably, it's a big philosophical shift from React Router, which relied more on absolute URLs. So, just rethink your URL usage a bit and it should all work A-OK.

**Comment by ryanflorence** [comment]
If we supported absolute relative URLS

1. It's not in line with how relative links work in the browser
2. You would lose absolute URLs completely.
3. It would be confusing looking at a link to know where it actually linked to, you'd need some context but you wouldn't *know* you needed context `<Link to="/somehwere"/>` vs. `<Link to ="../../some/where"/>`. The second communicates "you need to know where you are to know where this links to", the first you simply have no idea and you don't know that you need to know!

**Comment by Pyrolistical** [comment]
How do you implement a Home link that always goes to the basepath as a relative link?

If I'm on `/basepath/dashboard`, I can use `..` to go home. But if I'm on `/basepath/dashboard/admin` I need to do use `../..`. I have to track where I am?

What if I want a navbar that always points home? Then I would need to put basepath in two places.

**Comment by daviestar** [comment]
@Pyrolistical if you're happy with the trade-offs mentioned above, you could easily create this logic yourself, if it makes sense in your project:

```js
const BASE = '/client'

const App = () => (
  <Router basepath={BASE}>
    ...
  </Router>
)

const Link = ({ to = '', children, ... props }) => {
  const uri = to[0] === '/' ? BASE + to : to
  return (
    <reach.Link {... props} to={uri}>
      {children}
    </reach.Link>
  )
}
```

**Comment by astik** [comment]
@daviestar indeed, it is a good enough approach. We can do this wrapping for Link, navigate and Router. But i couldn't make it work for Redirect. Redirect can't be wrapped as it is expected as a direct child for reach Router.

Also, when I'm writing :
```
<Redirect from="/" to={`${BASE}/my/redirect`} noThrow />
```
It seems that the / into from attribute is not an absolute url (from a browser point of view), it is the base path defined for the router.

**Comment by benwiley4000** [comment]
@ryanflorence I disagree with "you would lose absolute URLs completely." I see what you're saying, but if the basepath defines the boundaries of my app, then any outside absolute urls would best be rendered with an `<a>` tag anyway, not a `<Link>`. IMO, this behavior is more confusing than it's worth and issue 3 that you mentioned doesn't seem like too big of a problem (to me).

Yes it's not in line with how links work for normal a tags... but a tags don't give you a wrapper element with a "basepath" either, and I'd guess this is how they *would* work if that were a thing.

**Comment by ColonelThirtyTwo** [comment]
I agree with @benwiley4000. You can't route above `basepath`, so there's no point in being able to `<Link>` there. You will never get a route match from within the app. It's useless.

Making `/` the root of the app isn't "confusing". The point of a clientside router is to route _within my app_, and `basepath` is what defines where my app is.

> 3\. It would be confusing looking at a link to know where it actually linked to, you'd need some context but you wouldn't _know_ you needed context `<Link to="/somehwere"/>` vs. `<Link to ="../../some/where"/>`. The second communicates "you need to know where you are to know where this links to", the first you simply have no idea and you don't know that you need to know!

Are you really arguing that an absolute URL needs more context than a relative URL? `<Link to="/somewhere/">` would link to `basepath + "/somewhere/"`, and since `basepath` is 99.99999% of the time a constant value, the URL is unambiguous. Meanwhile the relative URL is based on the current page you happen to be on.

**Comment by JiiB** [comment]
@ColonelThirtyTwo @benwiley4000 I had issues when I linked to a hash e.g. `<Link to="#scroll-top"`.

I forked the repo and changed the `resolve` function like this:
https://github.com/reach/router/blob/master/src/lib/utils.js#L149-L151

```js
  if (startsWith(to, "/") || startsWith(to, "#")) {
    return to;
  }
```

**Comment by will-moore** [comment]
I'm also struggling to understand why you would want ```<Link to="/">``` to link outside of the app. Surely that should always link to the ```basepath```, regardless of where you are.
Why would you really need context to know where ```<Link to="/somehwere"/>``` links to?
If I want an absolute link that ignores the React logic then I can use an ```<a>``` tag?

If this is different behaviour from ```React Router``` then then I hope that the ```React Router``` behaviour gets picked when the libraries are merged since I don't think I'm getting the "big philosophical shift" just yet. Which is a shame because in every other respect, ```Reach Router``` has been perfect.

**Comment by will-moore** [comment]
Thanks @daviestar for your workaround. I used it at https://github.com/will-moore/gallery2/commit/2e72ab2b57ab07fcc107c8c9a9b903022827f6f3

**Comment by thisissami** [comment]
Just chiming in here that the current behaviour is incredibly confusing. My expectation is that `basepath` sets the root of my app for all `Link`s and routes contained within. It took a while to understand that this is not actually a bug, but rather intended behaviour. 

If I wrote a PR to address this and change the behaviour to what many people in this thread clearly expect - would it be accepted, or rejected on the grounds that it would confused people?

**Comment by cevr** [comment]
This makes it necessary to add a custom Link component to handle base paths, especially for repos with multiple nested routers.

Like mentioned before, I think if we were using a link that routes outside the context of the basepath, then we should be using an anchor tag, rather than the context-sensitive Link component.

**Comment by izayl** [comment]
I think the `Link` Component can receive a `basepath` prop.
if `basepath` is given, the href should be relative to the basepath that setted in Router. 

Just like this:
```jsx

<Router basepath="/base/path">
    // pathname:  /base/path/a
    <Link basepath to="/a" />   

    // pathname:  /a
    <Link to="/a" />   
</Router>

```

**Comment by patspam** [comment]
Just wasted an hour on this - a note in the doc would be much appreciated by the next N people who run into this..

**Comment by tydira** [comment]
Oh bummer, this is still a thing. 🙃

**Comment by dvnrsn** [comment]
Is this not solved by `<Link to='./'>`. That's what seems to work for me.

**Comment by SimonFarrugia** [comment]
This is how I worked around this.
Extract the base pathname from the href of a hidden `<Link/>` next to your child `Router`, and expose the basepath through a hook:

Implementation:
```js
const BasepathContext = createContext("");

export function BaseRouter({ children, ...props }) {
    const [frag] = useState(document.createDocumentFragment());
    const [basepath, setBasepath] = useState("");

    useEffect(() => {
        // Get the base pathname from the href generated with @reach/router <Link/>
        const url = new URL(frag.firstChild.href);
        setBasepath(url.pathname);
    }, []);

    return (
      <>
        {createPortal(<Link to="" />, frag)}
        <BasepathContext.Provider value={basepath}>
          <Router {...props}>{children}</Router>
        </BasepathContext.Provider>
      </>
  );
}

export function useBasepath() {
    return useContext(BasepathContext) || "";
}
```

Then, use `<BaseRouter />` to define your routes, and build the Links inside your child components using the basepath from the useBasepath() hook.

Example:
```js
const basepath = useBasepath();

return <Link to={`${basepath}/whatever/it/may/be`}>My Link</Link>
```

**Comment by Pyrolistical** [comment]
not sure about that workaround implementation. i'm pretty sure there are easier ways to get the basepath.

Also, I think you can offer a even higher level api
```
const RelativeLink = useRelativeLink(();

return <RelativeLink to="whatever/it/may/be">My Link</Link>
```

and hide all the basepath stuff inside

**Comment by SimonFarrugia** [comment]
@Pyrolistical Please let me know if you find a cleaner way of getting the basepath. From what I remember when I looked at the reach router source, the `bauseUri` is on an internal `BaseContext` which is not accessible from outside. Maybe there is some reach router hook that we can harness instead, to simplify the implementation. Honestly I did not look into it further once the workaround with the `<Link />` worked :-).

BTW, there was an error in the implementation above. Forgot this line: `const url = new URL(frag.firstChild.href);`

Regarding the higher level Link component. Of course you can do that. I just wanted to provided the most raw and concise implementation so it is understandable how it works.

@ryanflorence Maybe we can have a built-in hook to get the baseuri/basepath from BaseContext?

**Comment by Pyrolistical** [comment]
@SimonFarrugia Just create your own router which wraps the existing one. So when the base path is passed to yours, you can capture it and then pass it along to the real router.

**Comment by SimonFarrugia** [comment]
@Pyrolistical Not sure I understand, maybe an example would help. How does "your router" get the base path if you cannot access reach router's internal context from the parent Router.

**Comment by Pyrolistical** [comment]
@SimonFarrugia 
```js
import {Router} from '@reach/router';

const BasepathContext = createContext("");

function BasepathCapturingRouter(props) {
  return <Router {...props}>
    <BasepathContext.Provider value={props.basepath}>
      {props.children}
    </BasepathContext.Provider>
  </Router>;
}
```

And then you `BasepathCapturingRouter` instead of `Router` directly

**Comment by SimonFarrugia** [comment]
@Pyrolistical But who passes the basepath in the props and where do you get it from? Maybe your use case is simpler. I have multiple Apps in different repos managed by different teams that maybe rendered in different places.

This may explain things better:
```js
const MainApp () => (
    <Router>
        <ChildApp path="one/*"/>
        <ChildApp path="abc/def/*"/>
    </Router>
);

const ChildApp = () => {
    // Inside BaseRouter, useBasepath() will return either "/one" or "/abc/def"
    // depending which route on the MainApp router is activated.
    return (
        <BaseRouter>
            // Inside any of the child routes below, no matter how deep, using useBasepath(),
            // you can create absolute Links instead of doing "../" or "../../" which may have to vary
            // depending how deep inside the ChildApp's routing tree the component is rendered.
            <MyFirstComponent path="first-route" />
            <MySecondComponent path="second-route" />
        </BaseRouter>
    );
};

const MyFirstComponent = () => {
    const basepath = useBasepath();

    return <>
        // Here to have a Link navigate to "second-route", I have to either use a relative path
        // which may have vary in in more complex use cases,
        <Link to="../second-route">My Link</Link>

        // or have an absolute path which works at every level
        <Link to={`${basepath}/second-route`}>My Link</Link>
    </>;
};
```

**Comment by Pyrolistical** [comment]
This has been fixed in react-router v6 see https://reactrouter.com/docs/en/v6/examples/multi-app

Just set the `BrowserRouter` `basename`
